sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java

```diff
@@ -196,119 +207,132 @@ public DruidSchema(
+                  final DruidTable oldTable = tables.put(dataSource, druidTable);
+                  if (oldTable == null || !oldTable.getRowSignature().equals(druidTable.getRowSignature())) {
+                    log.info("dataSource [%s] has new signature: %s.", dataSource, druidTable.getRowSignature());
+                  } else {
+                    log.debug("dataSource [%s] signature is unchanged.", dataSource);
                   }
                 }
+
+                initialized.countDown();
+              }
+              catch (InterruptedException e) {
+                // Fall through.
+                throw e;
+              }
+              catch (Exception e) {
+                log.warn(e, "Metadata refresh failed, trying again soon.");
+
+                synchronized (lock) {
+                  // Add our segments and dataSources back to their refresh and rebuild lists.
+                  segmentsNeedingRefresh.addAll(segmentsToRefresh);
+                  dataSourcesNeedingRebuild.addAll(dataSourcesToRebuild);
+                  lastFailure = System.currentTimeMillis();
+                  lock.notifyAll();
+                }
               }
-            }
-            catch (InterruptedException e) {
-              // Just exit.
-            }
-            catch (Throwable e) {
-              // Throwables that fall out to here (not caught by an inner try/catch) are potentially gnarly, like
-              // OOMEs. Anyway, let's just emit an alert and stop refreshing metadata.
-              log.makeAlert(e, "Metadata refresh failed permanently").emit();
-              throw e;
-            }
-            finally {
-              log.info("Metadata refresh stopped.");
             }
           }
+          catch (InterruptedException e) {
+            // Just exit.
+          }
+          catch (Throwable e) {
+            // Throwables that fall out to here (not caught by an inner try/catch) are potentially gnarly, like
+            // OOMEs. Anyway, let's just emit an alert and stop refreshing metadata.
+            log.makeAlert(e, "Metadata refresh failed permanently").emit();
+            throw e;
+          }
+          finally {
+            log.info("Metadata refresh stopped.");
+          }
+        }
+    );
+
+    ScheduledExecutors.scheduleWithFixedDelay(
+        localSegmentExec,
+        config.getMetadataRefreshPeriod().toStandardDuration(),
+        config.getMetadataRefreshPeriod().toStandardDuration(),
+        () -> {
+          synchronized (lock) {
+            // refresh known broadcast segments. Since DruidSchema is only present on the broker, any segment we have
+            // locally in the SegmentManager must be broadcast datasources. This could potentially be replaced in the
+            // future by fetching load rules from the coordinator
+            Set<String> localSegmentDatasources = segmentManager.getDataSourceNames();
+            dataSourcesNeedingRebuild.addAll(localSegmentDatasources);
```

Comment:
We don't really, I've refactored this pretty heavily.

I reworked the fix I did for #10017, to instead now preserve the segment/timeline event callbacks from the `BrokerServerView` (but still not add the server/segment to the timeline to avoid the weird loops), which in turn allows `DruidSchema` get these events for broker segments to mark datasources to be rebuilt in the normal path (just skipping metadata fetch on the assumption that the segment will appear, so that the existing loop handles these changes correctly.

Reply:
OK, I will take another look.